File: src/psk/pskapp/lib/assistant/conversation.dart
import 'assistant_models.dart';

/// Per-session state and pronoun resolution. "Take me there" only works if
/// there is a resolvable referent; with an empty focus stack we ask, never guess.
class ConversationSession {
  static const int maxFocus = 5;

  final List<FocusEntry> focusStack = [];
  PendingClarification? pendingClarification;
  PendingConfirmation? pendingConfirmation;

  void pushFocus(String entityId, String entityType, String name) {
    focusStack.removeWhere((f) => f.entityId == entityId);
    focusStack.insert(0, FocusEntry(entityId, entityType, name));
    if (focusStack.length > maxFocus) focusStack.removeRange(maxFocus, focusStack.length);
  }

  FocusEntry? resolveReference() => focusStack.isEmpty ? null : focusStack.first;

  void clearPending() {
    pendingClarification = null;
    pendingConfirmation = null;
  }

  void reset() {
    focusStack.clear();
    clearPending();
  }

  static const List<String> _referential = [
    'there', 'it', 'that', 'this one', 'that one', 'that game', 'this game',
    'take me there', 'open it', 'show me that', 'go there', 'launch that',
    'same one', 'the one', 'wahan', 'wahan le chalo', 'usko', 'wo wala', 'woh',
    'akkada', 'adi', 'daani',
  ];

  /// Word-boundary aware so "deposit" is not read as containing "it".
  static bool isReferential(String query) {
    final q = query.toLowerCase().trim().replaceAll(RegExp(r'[^\w\s]'), '');
    if (q.isEmpty || q.length > 40) return false;
    final words = q.split(RegExp(r'\s+')).where((w) => w.isNotEmpty).toSet();
    for (final phrase in _referential) {
      if (q == phrase) return true;
      if (phrase.contains(' ')) {
        final re = RegExp('\\b${phrase.replaceAll(RegExp(r'\s+'), r'\s+')}\\b');
        if (re.hasMatch(q)) return true;
      } else if (words.contains(phrase)) {
        return true;
      }
    }
    return false;
  }

  static const List<String> _affirm = ['yes', 'y', 'yeah', 'yep', 'ok', 'okay', 'sure', 'go', 'go ahead',
    'do it', 'confirm', 'haan', 'ha', 'theek hai', 'sari', 'avunu', 'please do'];
  static const List<String> _deny = ['no', 'n', 'nope', 'cancel', 'stop', 'wait', 'nahi', 'nahin', 'ledu',
    'not now', 'never mind', 'nevermind'];

  static String _clean(String q) => q.toLowerCase().trim().replaceAll(RegExp(r'[^a-z\s]'), '');

  static bool isAffirmation(String q) {
    final s = _clean(q);
    return _affirm.any((a) => s == a || s.startsWith('$a '));
  }

  static bool isDenial(String q) {
    final s = _clean(q);
    return _deny.any((d) => s == d || s.startsWith('$d '));
  }
}

class FocusEntry {
  final String entityId;
  final String entityType;
  final String name;
  FocusEntry(this.entityId, this.entityType, this.name);
}

class PendingClarification {
  final String query;
  final List<ClarificationOption> options;
  PendingClarification(this.query, this.options);
}

class PendingConfirmation {
  final String entityId;
  final String intent;
  PendingConfirmation(this.entityId, this.intent);
}

File: src/psk/pskwebsite/psk/server/src/assistant/conversation.js
/**
 * conversation.js — per-session state and pronoun resolution.
 *
 * "Take me there" only works if there is a resolvable referent. When the focus
 * stack is empty the correct behaviour is to ask, never to guess — guessing a
 * destination is how an assistant sends someone to the wrong financial page.
 *
 * In-memory with TTL. Swap for Redis or the sessions collection when you need
 * this to survive a restart or run multi-instance.
 */
const SESSIONS = new Map();
const TTL_MS = 30 * 60 * 1000;      // 30 minutes
const MAX_TURNS = 12;
const MAX_FOCUS = 5;

/** Phrases that refer back to something already mentioned. */
const REFERENTIAL = [
  'there', 'it', 'that', 'this one', 'that one', 'that game', 'this game',
  'take me there', 'open it', 'show me that', 'go there', 'launch that',
  'same one', 'the one', 'wahan', 'wahan le chalo', 'usko', 'wo wala', 'woh',
  'akkada', 'adi', 'daani',
];

function now() {
  return Date.now();
}

function prune() {
  const cutoff = now() - TTL_MS;
  for (const [id, s] of SESSIONS) {
    if (s.updatedAt < cutoff) SESSIONS.delete(id);
  }
}

export function getSession(sessionId) {
  prune();
  if (!SESSIONS.has(sessionId)) {
    SESSIONS.set(sessionId, {
      session_id: sessionId,
      turns: [],
      focus_stack: [],
      pending_clarification: null,
      pending_confirmation: null,
      createdAt: now(),
      updatedAt: now(),
    });
  }
  const s = SESSIONS.get(sessionId);
  s.updatedAt = now();
  return s;
}

export function addTurn(session, role, content, meta = {}) {
  session.turns.push({ role, content, at: new Date().toISOString(), ...meta });
  if (session.turns.length > MAX_TURNS) {
    session.turns = session.turns.slice(-MAX_TURNS);
  }
  session.updatedAt = now();
}

/** Push an entity to the front of the focus stack. */
export function pushFocus(session, entity) {
  if (!entity?.entity_id) return;
  session.focus_stack = [
    { entity_id: entity.entity_id, entity_type: entity.entity_type, name: entity.name,
      at: new Date().toISOString() },
    ...session.focus_stack.filter((f) => f.entity_id !== entity.entity_id),
  ].slice(0, MAX_FOCUS);
}

/**
 * Does the message rely on something previously mentioned?
 *
 * Matching is word-boundary aware on purpose: a naive substring check treats
 * "deposit" and "withdrawals" as containing the pronoun "it", which silently
 * hijacks ordinary questions into follow-ups against the focused entity.
 */
export function isReferential(query) {
  const q = String(query || '').toLowerCase().trim().replace(/[^\w\s]/g, '');
  if (!q) return false;
  if (q.length > 40) return false;                 // long messages carry their own subject

  const words = q.split(/\s+/).filter(Boolean);
  const wordSet = new Set(words);

  for (const phrase of REFERENTIAL) {
    if (q === phrase) return true;
    if (phrase.includes(' ')) {
      // multi-word phrase: match on a word-boundary regex
      const re = new RegExp(`\\b${phrase.replace(/\s+/g, '\\s+')}\\b`);
      if (re.test(q)) return true;
    } else if (wordSet.has(phrase)) {
      return true;
    }
  }
  return false;
}

/** The entity a pronoun refers to, or null. */
export function resolveReference(session) {
  return session.focus_stack[0] || null;
}

export function setPendingClarification(session, payload) {
  session.pending_clarification = payload;
}

export function setPendingConfirmation(session, payload) {
  session.pending_confirmation = payload;
}

export function clearPending(session) {
  session.pending_clarification = null;
  session.pending_confirmation = null;
}

const AFFIRM = ['yes', 'y', 'yeah', 'yep', 'ok', 'okay', 'sure', 'go', 'go ahead',
  'do it', 'confirm', 'haan', 'ha', 'theek hai', 'sari', 'avunu', 'please do'];
const DENY = ['no', 'n', 'nope', 'cancel', 'stop', 'wait', 'nahi', 'nahin', 'ledu',
  'not now', 'never mind', 'nevermind'];

export function isAffirmation(q) {
  const s = String(q || '').toLowerCase().trim().replace(/[^a-z\s]/g, '');
  return AFFIRM.includes(s) || AFFIRM.some((a) => s === a || s.startsWith(a + ' '));
}

export function isDenial(q) {
  const s = String(q || '').toLowerCase().trim().replace(/[^a-z\s]/g, '');
  return DENY.includes(s) || DENY.some((d) => s === d || s.startsWith(d + ' '));
}

export function sessionStats() {
  prune();
  return { active_sessions: SESSIONS.size };
}

export default {
  getSession, addTurn, pushFocus, isReferential, resolveReference,
  setPendingClarification, setPendingConfirmation, clearPending,
  isAffirmation, isDenial, sessionStats,
};
